# Introduction
## Exploring Hacker News Posts

Hacker News is a site started by the startup incubator Y Combinator, where user-submitted stories (known as "posts") are voted and commented upon, similar to reddit. Hacker News is extremely popular in technology and startup circles, and posts that make it to the top of Hacker News' listings can get hundreds of thousands of visitors as a result.

You can find the data set here, but note that it has been reduced from almost 300,000 rows to approximately 20,000 rows by removing all submissions that did not receive any comments, and then randomly sampling from the remaining submissions. Below are descriptions of the columns:

* `id:` The unique identifier from Hacker News for the post
* `title:` The title of the post
* `url:` The URL that the posts links to, if it the post has a URL
* `num_points:` The number of points the post acquired, calculated as the total number of upvotes minus the total number of downvotes
* `num_comments:` The number of comments that were made on the post
* `author:` The username of the person who submitted the post
* `created_at:` The date and time at which the post was submitted

We're specifically interested in posts whose titles begin with either Ask HN or Show HN. Users submit Ask HN posts to ask the Hacker News community a specific question. Below are a couple examples:

    Ask HN: How to improve my personal website?
    Ask HN: Am I the only one outraged by Twitter shutting down share counts?
    Ask HN: Aby recent changes to CSS that broke mobile?
    Ask HN: How to improve my personal website?
    Ask HN: Am I the only one outraged by Twitter shutting down share counts?
    Ask HN: Aby recent changes to CSS that broke mobile?
    
Likewise, users submit Show HN posts to show the Hacker News community a project, product, or just generally something interesting. Below are a couple of examples:

    Show HN: Wio Link  ESP8266 Based Web of Things Hardware Development Platform'
    Show HN: Something pointless I made
    Show HN: Shanhu.io, a programming playground powered by e8vm
    
We'll compare these two types of posts to determine the following:

* Do Ask HN or Show HN receive more comments on average?
* Do posts created at a certain time receive more comments on average?

Let's start by importing the libraries we need and reading the data set into a list of lists. we'll seperate the header row from rest of data at the begining so that we can analyze our data.

In [17]:
from csv import reader

### open Hacker News data set ###
opened_file = open("hacker_news.csv", encoding='utf8')
read_file = reader(opened_file)
hn = list(read_file)

# Separating Header from the list

hn_header = hn[0]
hn = hn[1:]
hn_header

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']

# Extracting Ask_Hn and Show_Hn Posts

Now we'll etract the posts starting with Ask_hn and show_hn as i'll make it easy to analyze further data.

In [18]:
ask_posts = []
show_posts = []
other_posts = []
for row in hn:
    title = row[1]
    if title.lower().startswith("ask hn"):
        ask_posts.append(row)
    elif title.lower().startswith("show hn"):
        show_posts.append(row)
    else:
        other_posts.append(row)
print("Length of Ask_posts:", len(ask_posts))
print("Length of Show_posts:", len(show_posts))
print("Length of Other_posts:", len(other_posts))

Length of Ask_posts: 1744
Length of Show_posts: 1162
Length of Other_posts: 17194


# Calculating Average no of Comments in each Post type

After getting post types separatel, let's find out avrage number of comments in ask_hn and show_hn.

In [19]:
total_ask_comments = 0
for row in ask_posts:
    
    total_ask_comments += int(row[4])
    avg_ask_comments = total_ask_comments/len(ask_posts)

print("Average Ask-Hn Comments:", avg_ask_comments)

Average Ask-Hn Comments: 14.038417431192661


In [20]:
total_show_comments = 0
for row in show_posts:
    
    total_show_comments += int(row[4])
    avg_show_comments = total_show_comments/len(show_posts)
    
print("Average Show-Hn Comments:", avg_show_comments)

Average Show-Hn Comments: 10.31669535283993


# Finding the Amount of Ask posts and comments by Hour created

We have seen that ask posts have more comments than show posts so further we'll analyze only ask posts

In [27]:
import datetime as dt
result_list = []
for row in ask_posts:
    created_at = row[6]
    no_of_comments = int(row[4])
    result_list.append([created_at, no_of_comments])
    
counts_by_hour = {}
comments_by_hour = {}
date_format = "%m/%d/%Y %H:%M"

for row in result_list:
    date = row[0]
    comment = row[1]
    time = dt.datetime.strptime(date, date_format).strftime("%H")
    if time not in counts_by_hour:
        counts_by_hour[time] = 1
        comments_by_hour[time] = comment
    else:
        counts_by_hour[time] += 1
        comments_by_hour[time] += comment
comments_by_hour

{'09': 251,
 '13': 1253,
 '10': 793,
 '14': 1416,
 '16': 1814,
 '23': 543,
 '12': 687,
 '17': 1146,
 '15': 4477,
 '21': 1745,
 '20': 1722,
 '02': 1381,
 '18': 1439,
 '03': 421,
 '05': 464,
 '19': 1188,
 '01': 683,
 '22': 479,
 '08': 492,
 '04': 337,
 '00': 447,
 '06': 397,
 '07': 267,
 '11': 641}

# Calculating the Average Number of Comments for Ask HN Posts by Hour

Now, we'll use these two dictionaries to calculate the average number of comments for posts created during each hour of the day.

In [29]:
avg_by_hour = []

for hour in comments_by_hour:
    avg_by_hour.append([hour, comments_by_hour[hour]/counts_by_hour[hour]])
avg_by_hour

[['09', 5.5777777777777775],
 ['13', 14.741176470588234],
 ['10', 13.440677966101696],
 ['14', 13.233644859813085],
 ['16', 16.796296296296298],
 ['23', 7.985294117647059],
 ['12', 9.41095890410959],
 ['17', 11.46],
 ['15', 38.5948275862069],
 ['21', 16.009174311926607],
 ['20', 21.525],
 ['02', 23.810344827586206],
 ['18', 13.20183486238532],
 ['03', 7.796296296296297],
 ['05', 10.08695652173913],
 ['19', 10.8],
 ['01', 11.383333333333333],
 ['22', 6.746478873239437],
 ['08', 10.25],
 ['04', 7.170212765957447],
 ['00', 8.127272727272727],
 ['06', 9.022727272727273],
 ['07', 7.852941176470588],
 ['11', 11.051724137931034]]

# Sorting and Printing Values from a List of Lists

Although we now have the results we need, this format makes it hard to identify the hours with the highest values. Let's finish by sorting the list of lists and printing the five highest values in a format that's easier to read.

In [35]:
swap_avg_by_hour = []
for elements in avg_by_hour:
    swap_avg_by_hour.append([elements[1], elements[0]])

print(swap_avg_by_hour, "\n")

sorted_swap = sorted(swap_avg_by_hour, reverse = True)
print(sorted_swap[:5])

[[5.5777777777777775, '09'], [14.741176470588234, '13'], [13.440677966101696, '10'], [13.233644859813085, '14'], [16.796296296296298, '16'], [7.985294117647059, '23'], [9.41095890410959, '12'], [11.46, '17'], [38.5948275862069, '15'], [16.009174311926607, '21'], [21.525, '20'], [23.810344827586206, '02'], [13.20183486238532, '18'], [7.796296296296297, '03'], [10.08695652173913, '05'], [10.8, '19'], [11.383333333333333, '01'], [6.746478873239437, '22'], [10.25, '08'], [7.170212765957447, '04'], [8.127272727272727, '00'], [9.022727272727273, '06'], [7.852941176470588, '07'], [11.051724137931034, '11']] 

[[38.5948275862069, '15'], [23.810344827586206, '02'], [21.525, '20'], [16.796296296296298, '16'], [16.009174311926607, '21']]


Let's make these findngs more readable.

In [38]:
print("Top 5 Hours for Ask-Hn Comments")
for avg, hr in sorted_swap[:5]:
    print(
        "{}: {:.2f} average comments per post".format(dt.datetime.strptime(hr, '%H').strftime("%H:%M"), avg)
)

Top 5 Hours for Ask-Hn Comments
15:00: 38.59 average comments per post
02:00: 23.81 average comments per post
20:00: 21.52 average comments per post
16:00: 16.80 average comments per post
21:00: 16.01 average comments per post


The hour that receives the most comments per post on average is 15:00, with an average of 38.59 comments per post. There's about a 60% increase in the number of comments between the hours with the highest and second highest average number of comments.

According to the data set [documentation](https://www.kaggle.com/hacker-news/hacker-news-posts), the timezone used is Eastern Time in the US. So, we could also write 15:00 as 3:00 pm est.

# Conclusion

In this project, we analyzed ask posts and show posts from Hacker News to determine which type of post and time receive the most comments on average. Based on our analysis, to maximize the amount of comments a post receives, we'd recommend the post be categorized as ask post and created between 15:00 and 16:00 (3:00 pm est - 4:00 pm est).

However, it should be noted that the data set we analyzed excluded posts without any comments. Given that, it's more accurate to say that of the posts that received comments, ask posts received more comments on average and ask posts created between 15:00 and 16:00 (3:00 pm est - 4:00 pm est) received the most comments on average.